cpp/DEVELOPER_GUIDE.md

```diff
@@ -11,6 +11,26 @@ The implementation of cuML is single threaded.
 
 ## Coding style
 
+## Code format
+### Introduction
+Coding format checker is a way to enforce uniform coding format across all C++/CUDA sources in cuML repo. To achieve this, we are relying on `clang-format`. And the coding style we are following is based on google style guide [here](https://google.github.io/styleguide/cppguide.html#Formatting). The only digressions from this style are:
+1. Do not split empty functions/records/namespaces
+2. Keep the indentation as 2 spaces "everywhere", including the cases of continuation lines.
+3. Disable reflowing of comments
+To know the reasons behind these deviations from the google style guide, refer to the comments [here](./.clang-format).
+
+### How is the check done?
+[run-clang-format.py](scripts/run-clang-format.py) is run as a first thing when a developer runs `make`. This script figures out the set of files that the developer has modified/staged and runs clang-format on them. In case there are some formatting differences between the one done by the dev and the one suggested by clang-format, an error is raised and the build fails.
+
+### How to know the formatting violations?
+When there are formatting errors, the above script prints a 'diff' command, using which one could figure out the places where there are formatting differences. Unfortunately, unlike `flake8`, `clang-format` does NOT print descriptions of the violations itself, but instead directly formats the code. So, the only way currently to know why there are formatting differences is to run the diff command as suggested by this script against each violating source file.
```

Comment:
```suggestion
When there are formatting errors, `run-clang-format.py` prints a `diff` command, showing where there are formatting differences. Unfortunately, unlike `flake8`, `clang-format` does NOT print descriptions of the violations, but instead directly formats the code. So, the only way currently to know why there are formatting differences is to run the diff command as suggested by this script against each violating source file.
```

Reply:
FWIW, this kind of sucks...

Reply:
Agreed Mark. After using this feature for a while, maybe it's best to file an issue an clang-tooling itself.